content/src/content/jcr_root/apps/core/wcm/components/commons/datalayer/clientlibs/js/datalayer.js

```diff
@@ -0,0 +1,218 @@
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ ~ Copyright 2019 Adobe
+ ~
+ ~ Licensed under the Apache License, Version 2.0 (the "License");
+ ~ you may not use this file except in compliance with the License.
+ ~ You may obtain a copy of the License at
+ ~
+ ~     http://www.apache.org/licenses/LICENSE-2.0
+ ~
+ ~ Unless required by applicable law or agreed to in writing, software
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+(function() {
+    "use strict";
+
+    /* eslint no-console: "off" */
+
+    var CHANGE_EVENT = "datalayer:change";
+
+    // Initializes the data layer
+    function init() {
+        window.dataLayer = window.dataLayer || [];
+        window.dataLayer.state = {};
+        window.dataLayer._listeners = [];
+        populateDataLayer();
+        handleEvents(window.dataLayer);
+        overridePush();
+        populateDataLayerAfterOverride();
+        console.log("data layer script initialized");
+    }
+
+    // Augments the push function (only for the data layer object) to also handle the event
+    function overridePush() {
+        window.dataLayer.push = function() {
+            var pushArguments = arguments;
+            Object.keys(pushArguments).forEach(function(key) {
+                handleEvent(pushArguments[key]);
+            });
+            return Array.prototype.push.apply(this, pushArguments);
+        };
+    }
+
+    function handleEvents(dataLayer) {
+        dataLayer.forEach(function(event) {
+            handleEvent(event);
```

Comment:
we should remove listeners from the array when we initially scrape the events that are defined before the script has initialized.